File: Racional.java
public class Racional {
    private int numerador;
    private int denominador;

    Racional(int numerador, int denominador) {
        // validação de primos entre si
        for (int i = Math.min(numerador, denominador); i >= 2; i--) {
            if (numerador % i == 0 && denominador % i == 0) {
                numerador = numerador / i;
                denominador = denominador / i;

            }
        }

        this.numerador = numerador;
        this.denominador = denominador;

    }

    public double razao() {
        return (double) numerador / denominador;
    }

    public Racional sum(Racional r1) {
        int numerador = (r1.numerador * this.denominador) + (this.numerador * r1.denominador);
        int denominador = this.numerador * this.denominador;

        return new Racional(numerador, denominador);
        
    }

    public Racional sub(Racional r1) {
        int numerador = (r1.numerador * this.denominador) - (this.numerador * r1.denominador);
        int denominador = this.numerador * this.denominador;
        return new Racional(numerador, denominador);
    }

    public Racional mult(Racional r1) {
        int numerador = this.numerador * r1.numerador;
        int denominador = this.denominador * r1.denominador;
        return new Racional(numerador, denominador);
    }

}

File: racional.py
class Racional:    
    def __init__(self, __numerador: int, __denominador: int) -> None:
        
        #validação para garantir que são primos entre si
        for i in range(min(__numerador, __denominador), 1, -1):
            if __numerador % i == 0 and __denominador % i == 0:
                __numerador //= i
                __denominador //= i
                break
            
        
        self.__numerador = __numerador
        self.__denominador = __denominador
    
    def __str__(self) -> str:
        return f"{self.__numerador}/{self.__denominador}"
    
    def razao(self) -> float:
        return self.__numerador / self.__denominador
    
    def sum(self, r1) -> "Racional":
        numerador = (r1.__numerador * self.__denominador) + (self.__numerador * r1.__denominador)
        denominador = self.__numerador * self.__denominador
        return Racional(numerador, denominador)
    
    def sub(self, r1) -> "Racional":
        numerador = (r1.__numerador * self.__denominador) - (self.__numerador * r1.__denominador)
        denominador = self.__numerador * self.__denominador
        return Racional(numerador, denominador)
    
    def mult(self, r1) -> "Racional":
        numerador = self.__numerador * r1.__numerador
        denominador = self.__denominador * r1.__denominador
        return Racional(numerador, denominador)
    

        